src/SignalR/server/Core/src/Internal/ChannelExtensions.cs

```diff
@@ -0,0 +1,36 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Microsoft.AspNetCore.SignalR.Internal;
+
+internal static class ChannelExtensions
+{
+    public static ValueTask RunAsync<TState>(this ChannelBasedSemaphore channelSemaphore, Func<TState, Task> callback, TState state)
+    {
+        if (channelSemaphore.AttemptWait())
+        {
+            _ = RunTask(callback, channelSemaphore, state);
+            return ValueTask.CompletedTask;
+        }
+
+        return RunSlowAsync(channelSemaphore, callback, state);
+    }
+
+    private static async ValueTask RunSlowAsync<TState>(this ChannelBasedSemaphore channelSemaphore, Func<TState, Task> callback, TState state)
+    {
+        _ = await channelSemaphore.WaitAsync();
+        _ = RunTask(callback, channelSemaphore, state);
+    }
+
+    static async Task RunTask<TState>(Func<TState, Task> callback, ChannelBasedSemaphore channelSemaphore, TState state)
+    {
+        try
+        {
+            await callback(state);
+        }
+        finally
+        {
+            await channelSemaphore.ReleaseAsync();
```

Comment:
I do not understand why we would ever await a release. It should not be possible for the semaphore to ever go over its max capacity, right? I thought we were only releasing the semaphore if we had already acquired it.

Reply:
Other invocations can acquire the semaphore in the meantime. This is why I previously had removed the semaphore max (or double it when we discussed offline), but for channels we can instead await the "release" which basically means wait for any invocations to complete, or for any invocations to call `InvokeAsync` which unblocks the channel.

Reply:
Ah, I think if you `_ = Task.Run(async () => _ = await caller.InvokeAsync<int>(...));` in a hub method then this wouldn't be able to be released synchronously. Or just `_ = caller.InvokeAsync<int>(...);` And if we changed it to TryWrite and throw on failure then the connection would be closed.

Reply:
This seems sketchy. If the corresponding hub invocation has already completed, we have no reason to be calling `ReleaseAsync()` in `InvokeAsync()`, we should just skip that in that case. It's not blocking the receive loop.

The corollary is that if `InvokeAsync()` has already released the semaphore before the hub invocation completes, we should not release it again here. I don't think we should ever release the semaphore more than once for a given hub invocation.

The reason this is important is that if someone does have N ongoing *background* invocations, the current logic allows N extra non-client-invoking hub methods to run concurrently. That makes no sense to me when that can be avoided by ensuring there's at most only one call to Release for every hub method invoked.

I think it's debatable whether the client proxies should even be usable after you've returned from your hub method. But give that's allowed, we need to synchronize so it's never possible for background calls to `InvokeAsync()` to magically increase the `MaximumParallelInvocationsPerClient` for non-streaming, non-client-invoking hub methods.